### ISSUE TITLE
完型填空的例子能正确运行，导出推理模型不报错，但是推理模型为无效模型。

### ISSUE THREAD

**Issue by hello-web** [PaddlePaddle/ERNIE#532]
# coding=gbk

import os
import re
import time
import logging
import json
from random import random
from tqdm import tqdm
from functools import reduce, partial

import numpy as np
import logging
import argparse

import paddle
import paddle.fluid as F
import paddle.fluid.dygraph as FD
import paddle.fluid.layers as L

from propeller import log
import propeller.paddle as propeller

log.setLevel(logging.DEBUG)
logging.getLogger().setLevel(logging.DEBUG)

#from model.bert import BertConfig, BertModelLayer
#from ernie.modeling_ernie import ErnieModel, ErnieModelForSequenceClassification
from ernie.tokenizing_ernie import ErnieTokenizer, ErnieTinyTokenizer
from ernie.optimization import AdamW, LinearDecay
from ernie.modeling_ernie import ErnieModelForPretraining, ErnieModel

if __name__ == '__main__':
    parser = argparse.ArgumentParser('save_inference_model with ERNIE')
    parser.add_argument('--from_pretrained', type=str, required=True, help='pretrained model directory or tag')
    parser.add_argument('--inference_model_dir', type=str, default=None, help='inference model output directory')

    args = parser.parse_args()

    tokenizer = ErnieTokenizer.from_pretrained(args.from_pretrained)
    rev_dict = {v: k for k, v in tokenizer.vocab.items()}
    rev_dict[tokenizer.pad_id] = '' # replace [PAD]
    rev_dict[tokenizer.sep_id] = '' # replace [PAD]
    rev_dict[tokenizer.unk_id] = '' # replace [PAD]

    @np.vectorize
    def rev_lookup(i):
        return rev_dict[i]

    place = F.CPUPlace()
    with FD.guard(place):
        model = ErnieModelForPretraining.from_pretrained(args.from_pretrained)

        src, _ = tokenizer.encode('戊戌变法，又称百日维新，是 [MASK] [MASK] [MASK] 、梁启超等维新派人士通过光绪帝进行 的一场资产阶级改良。')
        print(src)
        src_ids = np.expand_dims(src, 0)
        src_ids = FD.to_variable(src_ids)

        if args.inference_model_dir is not None:
            log.debug('saving inference model')
            class InferemceModel(ErnieModelForPretraining):
                def __init__(self, *args, **kwargs):
                    super(InferemceModel, self).__init__(*args, **kwargs)
                    #del self.pooler_heads
                def forward(self, src_ids, *args, **kwargs):
                    #mlm_pos = kwargs.pop('mlm_pos', np.int64(3))
                    pooled, encoded = ErnieModel.forward(self, src_ids, *args, **kwargs)
                    encoded_2d = L.gather_nd(encoded, L.where(src_ids == 3))
                    encoded_2d = self.mlm(encoded_2d)
                    encoded_2d = self.mlm_ln(encoded_2d)
                    logits_2d = L.matmul(encoded_2d, self.word_emb.weight, transpose_y=True) + self.mlm_bias
                    return logits_2d

            model.__class__ = InferemceModel 
            logits = model(src_ids)
            _, static_model = FD.TracedLayer.trace(model, inputs=[src_ids])
            static_model.save_inference_model(args.inference_model_dir)
            
            logits = logits.numpy()
            output_ids = np.argmax(logits, -1)
            seg_txt = rev_lookup(output_ids)
            print(''.join(seg_txt))

以上为复现代码
问题：完型填空的例子能正确运行，导出推理模型不报错，但是推理模型为无效模型，不能进行正确推理。
导出的模型文件参考文件名：
__model__
create_parameter_0.b_0
embedding_[0-2].w_0
layer_norm_[0-25].b_0
layer_norm_[0-25].w_0
linear_[0-74].b_0
linear_[0-74].w_0

**Comment by kirayummy** [comment]
save_inference_model需要使用inference代码执行，这个项目可采用dygraph的save接口保存模型文件，见
https://www.paddlepaddle.org.cn/documentation/docs/zh/api_cn/dygraph_cn/save_dygraph_cn.html

**Comment by hello-web** [comment]
> save_inference_model需要使用推理代码执行，这个项目可采用dygraph的保存接口保存模型文件，见
> https://www.paddlepaddle.org.cn/documentation/docs/zh/api_cn/dygraph_cn/save_dygraph_cn.html

参考 demo 里的 finetune_classifier_dygraph.py 推理保存接口 save_inference_model 完成上面的推理模型导出实现不了。原始 demo 可以在不训练、不预测的情况下加载预训练模型后正常导出。

https://github.com/PaddlePaddle/ERNIE/blob/4eb65ce4ff79028c44f6441ad19ea9aab664b434/demo/finetune_classifier_dygraph.py#L152

**Comment by kirayummy** [comment]
替换保存接口试试
```
# save
FD.save_dygraph(model.state_dict(), args.inference_model_dir)

# load
m, _ = FD.load_dygraph(args.inference_model_dir)
model.set_dict(m)

```

**Comment by hello-web** [comment]
替换保存接口和目标不一样
替换保存接口后保存的是动态图检测点，和保存推理模型的目标不一致。

初步判断是生成策略问题：
windows 下复现应该是 ngraph_subgraph_pass 的问题。
linux 下，在 AI Studio 中复现，应该是 fusion_group_pass 的问题。

**Comment by hello-web** [comment]
AI Studio 中复现
https://aistudio.baidu.com/aistudio/projectdetail/671497
生成的错误推理模型在 ./data/out 目录
